### PR TITLE
Update Developer's Guide

### PR DESCRIPTION
**Pull request by pyokagan**
TODO:
* [ ] undo/redo
* [x] OverloadParser / SubcommandParser
* [x] Update `Model` to include `ReadOnlyModel`.
* [x] How `HelpCommandResult`, `ExitCommandResult` works
* [x] Fix "setdatadir" autocomplete bug :X (Or remove it)

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -407,8 +374,145 @@ interface and saves/loads `ReadOnlyTaskBook`s.
...
+5. If the `Model` has been modified as a result of the command, `Logic` will
+   then write the updated `Model` back to disk using the `Storage` component.
+
+Given in Figure 2.5 below is the sequence diagram for interactions within the
+`Logic` component for the `execute("delete 1")` API call.
+
+<figure>
+<img src="images/devguide/seq-deleteevent.png">
+<figcaption><div align="center">Figure 2.5: Sequence diagram for event deletion</div></figcaption>
+</figure>
```
Comment (kychua) [inline docs/DeveloperGuide.md:466]: - `cmd` is calling `addEventTask()`? 
- arrow from `:Storage` to `:Logic` should be dotted

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -407,8 +374,145 @@ interface and saves/loads `ReadOnlyTaskBook`s.
...
+One major advantage of this approach is that it is extremely easy to mix and
+match parsers to generate create new command formats, since each piece of
+parsing logic is neatly encapsulated into individual classes. It is also
+extremely easy to unit test each individual parser. Another major advantage of
+this approach is that it is very DRY -- each piece of parsing logic has a
+single definitive source. For example, if we wanted to modify the task index
+format such that instead of `e1`, `d1`, `f1`, users have to type `Banana1`,
+`Apple1`, `Papaya1`, all we need to do is to modify the `IndexParser`, and all
+commands will immediately recognise the new format.
+
```
Comment (kychua) [inline docs/DeveloperGuide.md]: would probably be useful if it explains from the new developer's perspective i.e. what exactly the new developer needs to do if he/she wants to add a new command format to parse & use this to explain the design e.g. what common parsers are available for use? (IndexParser, DateParser, CommandLineParser etc.?), need to make sure it implements Parser interface, what are Flag, Argument etc

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -286,6 +243,11 @@ used to manage the logging levels and logging destinations.
 
 ### Model implementation
 
+<figure>
+<img src="images/devguide/comp-model.png">
+<figcaption><div align="center">Figure 2.2: Model component class diagram</div></figcaption>
+</figure>
+
```
Comment (kychua) [inline docs/DeveloperGuide.md:245]: What is `WorkingItem` for?

**Review by kychua — COMMENTED** [review]
- better to move this down: 
  
- can also discuss undo/redo implementation & interesting design decisions or design patterns used

**Comment by pyokagan** [comment]
Ugh, really need to complete this by Sunday ><

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (pyokagan) [reply docs/DeveloperGuide.md:466]: > cmd is calling addEventTask()?

:O whoops

**Pull request merged**